test/tools/runner/index.js

```diff
@@ -18,92 +17,84 @@ const MONGODB_API_VERSION = process.env.MONGODB_API_VERSION;
 const SINGLE_MONGOS_LB_URI = process.env.SINGLE_MONGOS_LB_URI;
 // Load balancer fronting 2 mongoses.
 const MULTI_MONGOS_LB_URI = process.env.MULTI_MONGOS_LB_URI;
+const loadBalanced = SINGLE_MONGOS_LB_URI && MULTI_MONGOS_LB_URI;
 const filters = [];
 
-function initializeFilters(client, callback) {
+const LOG_FILTER_REASON = false;
+
+let initializedFilters = false;
+async function initializeFilters(client) {
+  if (initializedFilters) {
+    return;
+  }
+  initializedFilters = true;
+  const context = {};
+
   const filterFiles = fs
     .readdirSync(path.join(__dirname, 'filters'))
     .filter(x => x.indexOf('js') !== -1);
 
-  // context object that can be appended to as part of filter initialization
-  const context = {};
+  for (const filterName of filterFiles) {
+    const FilterModule = require(path.join(__dirname, 'filters', filterName));
+    const filter = new FilterModule();
 
-  eachAsync(
-    filterFiles,
-    (filterName, cb) => {
-      const FilterModule = require(path.join(__dirname, 'filters', filterName));
-      const filter = new FilterModule();
+    console.assert(typeof filter === 'object');
+    console.assert(filter.filter && typeof filter.filter === 'function');
 
-      if (typeof filter !== 'object') {
-        cb(new TypeError('Type of filter must be an object'));
-        return;
-      }
+    filters.push(filter);
 
-      if (!filter.filter || typeof filter.filter !== 'function') {
-        cb(new TypeError('Object filters must have a function named filter'));
-        return;
-      }
+    if (typeof filter.initializeFilter === 'function') {
+      await new Promise((resolve, reject) =>
+        filter.initializeFilter(client, context, e => (e ? reject(e) : resolve()))
+      );
+    }
+  }
 
-      filters.push(filter);
-      if (typeof filter.initializeFilter === 'function') {
-        filter.initializeFilter(client, context, cb);
-      } else {
-        cb();
-      }
-    },
-    err => callback(err, context)
-  );
+  return context;
 }
 
-function filterOutTests(suite) {
-  suite.tests = suite.tests.filter(test => filters.every(f => f.filter(test)));
-  suite.suites.forEach(suite => filterOutTests(suite));
-}
+beforeEach(async function () {
+  if (Object.keys(this.currentTest.metadata).length > 0) {
+    let ok = true;
+    for (const filter of filters) {
+      ok = ok && filter.filter(this.currentTest);
```

Comment:
just a sanity check - is the metadata prop always defined?

since the only way for the test to be skipped is for a filter to get applied, we can simplify the nesting here, I think, no need to keep track of an "ok", just do something like 
```js
const failedFilter = filters.find(filter => !filter.filter(this.currentTest));
if (failedFilter) { 
  // LOG_FILTER_REASON logic
  this.skip();
}
```

Reply:
Cleaned this up, much more streamlined thanks!
cc: @baileympearson pinging you for another look cus of the changes I mention below:
Annnnd, modified the filter logging logic, I was using a conditional hack here of modifying the test name so that you could see it for debugging purposes but its useful information in our daily workflow I think. Apparently Go's built in test framework has a way of printing the skip reason for a test so I gave it a shot for us, here's what it looks like:

```
Retryable Reads
    aggregate-serverErrors - single,replicaset
      ✔ Aggregate fails after NotWritablePrimary when retryReads is false
    aggregate-serverErrors - sharded,load-balanced
      - Aggregate fails after NotWritablePrimary when retryReads is false
        ↬ filtered by MongoDBTopologyFilter - { requires: { topology: [ 'sharded', 'load-balanced' ], mongodb: '>= 4.1.7', authEnabled: false, serverless: undefined } }
``` 

We could even use this to attach reasons for our manual skips, checkout `monitoring.test.js` The ticket number will now get logged next to the test. I think now that mocha will log all test names on all runs, its an easy way for folks to find common spec tests and quickly locate the reason for a skip! And our xunit output file will still have only the test name so the test history is preserved in evergreen! 

shout out to @divjotarora for giving me the idea! 🚀